File: head_first_design_pattern/Python/adapter.py
from abc import ABC, abstractmethod


class Duck(ABC):
    def quack(self):
        pass

    def fly(self):
        pass


class MallardDuck(Duck):
    def quack(self):
        print("Quack")

    def fly(self):
        print("I'm flying")


class Turkey(ABC):
    def gobble(self):
        pass

    def fly(self):
        pass


class WildTurkey(Turkey):
    def gobble(self):
        print("Gobble google")

    def fly(self):
        print("I'm flying a short distance")

#object adaptive
class TurkeyAdapter(Duck):
    def __init__(self, turkey):
        self.turkey = turkey

    def quack(self):
        self.turkey.gobble()

    def fly(self):
        for i in range(0, 5):
            self.turkey.fly()


def test_duck(duck):
    duck.quack()
    duck.fly()

duck = MallardDuck()
turkey = WildTurkey()
turkeyAdapter = TurkeyAdapter(turkey)

print("The Turkey says...")
turkey.gobble()
turkey.fly()

print("The Duck says...")
test_duck(duck)

print("The TurkeyAdapter says..")
test_duck(turkeyAdapter)

File: head_first_design_pattern/C++/adapter.cpp
#include"adapter.h"
#include<iostream>

using namespace std;

void MallardDuck::quack() {
	cout <<"Quack"<< endl;
}

void MallardDuck::fly() {
	cout << "I'm flying" << endl;
}

void WildTurkey::gobble() {
	cout <<"Gobble google"<< endl;
}

void WildTurkey::fly() {
	cout <<"I'm flying a short distance"<< endl;
}

TurkeyAdapter::TurkeyAdapter(Turkey* turkey) {
	this->turkey = turkey;
};

void TurkeyAdapter::quack() {
	this->turkey->gobble();
};

void TurkeyAdapter::fly() {
	this->turkey->fly();
}

void Turkey::gobble() {
	cout << "In" << endl;
};
void Turkey::fly() {};